Crash with Turn Off the Lights extension enabled

kiaraarose opened the pull request:
#### b5ff6a931f2d3000b5b6fc1b4178ac5c77ce950a
<pre>
Crash with Turn Off the Lights extension enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=269924">https://bugs.webkit.org/show_bug.cgi?id=269924</a>
<a href="https://rdar.apple.com/123323719">rdar://123323719</a>

Reviewed by Timothy Hatcher.

This assertion is being hit:
*** Assertion failure in -[_UIMenuBuilder _insertMenu:intoParentMenu:newParentMenu:], UIMenuBuilder.m:443

This is because &quot;duplicate&quot; commands are being inserting into the menu. Per UIKit,
commands cannot have the same selector if their propertyList is the same. To fix this,
pass in a propertyList to each command so they can be distinguished from one other.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:handler:]):

Canonical link: <a href="https://commits.webkit.org/275193@main">https://commits.webkit.org/275193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74fe963969d1270321ce1f585ef336c923837f0b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20151 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43516 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43700 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/37230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17482 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41712 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/17078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/35432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45018 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/37333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/17572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/17624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5484 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
```diff
@@ -54,7 +58,12 @@ + (instancetype)commandWithTitle:(NSString *)title image:(UIImage *)image input:
+    auto *propertyList = @{
+        titleKey: title,
+        activationKey: input,
+};
```
Indent.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
```diff
@@ -46,6 +46,10 @@
 #if PLATFORM(IOS_FAMILY)
+
+static NSString * const titleKey = @"title";
+static NSString * const activationKey = @"activationKey";
```
I would just inline these string literals in the dictionary, they are not used anywhere else.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
```diff
@@ -46,6 +46,7 @@
 #endif
 
 #if PLATFORM(IOS_FAMILY)
+
```
remove

kiaraarose committed:
Crash with Turn Off the Lights extension enabled
https://bugs.webkit.org/show_bug.cgi?id=269924
rdar://123323719

Reviewed by Timothy Hatcher.

This assertion is being hit:
*** Assertion failure in -[_UIMenuBuilder _insertMenu:intoParentMenu:newParentMenu:], UIMenuBuilder.m:443

This is because "duplicate" commands are being inserting into the menu. Per UIKit,
commands cannot have the same selector if their propertyList is the same. To fix this,
pass in a propertyList to each command so they can be distinguished from one other.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:handler:]):

Canonical link: https://commits.webkit.org/275193@main
